aten/src/ATen/native/cpu/FlashAttentionKernel.cpp

```diff
@@ -306,16 +423,136 @@ void cpu_flash_attention(
   accum_t* buf_data = buf.data_ptr<accum_t>();
   scalar_t* buf_reduced_data = is_reduced_type ? buf_reduced.data_ptr<scalar_t>() : nullptr;
 
+  // Buffer to store padding query
+  scalar_t* query_padding_ptr = nullptr;
+  std::unique_ptr<unsigned short[]> query_padding_data;
+  if (!headSize_even && need_pack) {
+    query_padding_data =
+        std::make_unique<unsigned short[]>(num_thread * qSplitSize * eheadSize);
+    query_padding_ptr = reinterpret_cast<scalar_t*>(query_padding_data.get());
+  }
+  // Buffer to store Key and Value after transforms
+  at::Tensor key_t_reorder = at::empty(
+      {batchSize,
+       num_head,
+       eheadSize,
+       rkvSize},
+      c10::CppTypeToScalarType<scalar_t>::value);
+  auto key_reorder_ptr = key_t_reorder.data_ptr<scalar_t>();
+
+  scalar_t* value_reorder_ptr = nullptr;
+  std::unique_ptr<unsigned short[]> value_reorder_data;
+  int kv_padding_size = (kvSize - 1) / kvSplitSize * ekvSplitSize + ekvTail;
+  if (need_pack) {
+    value_reorder_data = std::make_unique<unsigned short[]>(
+        batchSize * num_head * kv_padding_size * rHeadSize);
+    value_reorder_ptr = reinterpret_cast<scalar_t*>(value_reorder_data.get());
+  }
+
+  // Reorder K, V
+  if (need_pack) {
+    at::parallel_for(0, batchSize * num_head * kvSlice, 1, [&](int64_t begin, int64_t end) {
+        int64_t i = 0, j = 0, l = 0, n = 0;
+        at::native::data_index_init(begin, i, batchSize, j, num_head, l, kvSlice);
+        std::unique_ptr<unsigned short[]> transpose_buffer =
+            std::make_unique<unsigned short[]>(eheadSize * packb_size);
+        scalar_t* transpose_buffer_ptr = reinterpret_cast<scalar_t*>(transpose_buffer.get());
+        std::unique_ptr<unsigned short[]> v_copy_buffer =
+            std::make_unique<unsigned short[]>(ekvSplitSize * packb_size);
+        scalar_t* v_copy_buffer_ptr = reinterpret_cast<scalar_t*>(v_copy_buffer.get());
+        for (const auto z : c10::irange(begin, end)) {
+          (void)z; // Suppress unused variable
+          n = l * kvSplitSize;
+          int64_t kvBlockSize = std::min(kvSplitSize, kvSize - n);
+          int64_t ekvBlockSize = kvBlockSize % 2 == 0 ? kvBlockSize : kvBlockSize + 1;
+
+          // Split kvSplitSize with packb_size
+          // [kvSplitSize, headSize] -> [div_up(kvSplitSize, packb_size), packb_size, headSize]
+          // Transpose [packb_size, headSize] -> [headSize, packb_size]
+          // Pack transposed buffer
+
+          for (int64_t b = 0; b < kvBlockSize; b += packb_size) {
+            bool tail = kvBlockSize - b < packb_size;
+            // TODO Use transpose pack when it is supported
```

Comment:
I guess you mean "fuse transpose and pack" here?

Reply:
Yes